modules/phase_field/src/userobjects/PolycrystalEBSD.C

```diff
@@ -0,0 +1,98 @@
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+
+#include "PolycrystalEBSD.h"
+#include "EBSDReader.h"
+
+template <>
+InputParameters
+validParams<PolycrystalEBSD>()
+{
+  InputParameters params = validParams<PolycrystalUserObjectBase>();
+  params.addClassDescription("Object for setting up a polycrystal structure from an EBSD Datafile");
+  params.addParam<UserObjectName>("ebsd_reader", "EBSD Reader for initial condition");
+  params.addParam<unsigned int>("phase", "EBSD phase number from which to retrieve information");
+  return params;
+}
+
+PolycrystalEBSD::PolycrystalEBSD(const InputParameters & parameters)
+  : PolycrystalUserObjectBase(parameters),
+    _ebsd_reader(getUserObject<EBSDReader>("ebsd_reader")),
+    _consider_phase(isParamValid("phase")),
+    _phase(isParamValid("phase") ? getParam<unsigned int>("phase") : 0),
+    _node_to_grain_weight_map(_ebsd_reader.getNodeToGrainWeightMap())
+{
+}
+
+void
+PolycrystalEBSD::getGrainsBasedOnPoint(const Point & point,
+                                       std::vector<unsigned int> & grains) const
+{
+  const EBSDAccessFunctors::EBSDPointData & d = _ebsd_reader.getData(point);
+  const auto phase = d._phase;
+
+  // See if we are in a phase that we are actually tracking
+  if (_consider_phase && phase != _phase)
+  {
+    grains.resize(0);
+    return;
+  }
+
+  // Get the ids from the EBSD reader
+  const auto global_id = _ebsd_reader.getGlobalID(d._feature_id);
+  const auto local_id = _ebsd_reader.getAvgData(global_id)._local_id;
+
+  grains.resize(1);
+  grains[0] = _consider_phase ? local_id : global_id;
+}
+
+unsigned int
+PolycrystalEBSD::getNumGrains() const
+{
+  return _ebsd_reader.getGrainNum();
+}
+
+Real
+PolycrystalEBSD::getVariableValue(unsigned int op_index, const Node & n) const
+{
+  // Make sure the _current_node is in the node_to_grain_weight_map (return error if not in map)
+  const auto it = _node_to_grain_weight_map.find(n.id());
+
+  if (it == _node_to_grain_weight_map.end())
+    mooseError("The following node id is not in the node map: ", n.id());
+
+  // Increment through all grains at node_index (these are global IDs if consider_phase is false and
+  // local IDs otherwise)
+  const auto num_grains = getNumGrains();
+  for (auto index = decltype(num_grains)(0); index < num_grains; ++index)
+  {
+    // If the current order parameter index (_op_index) is equal to the assigned index
+    // (_assigned_op),
+    // set the value from node_to_grain_weight_map
+    auto value = (it->second)[_consider_phase ? _ebsd_reader.getGlobalID(_phase, index) : index];
+    if (_grain_to_op[index] == op_index && value > 0.0)
+      return value;
+  }
+
+  return 0.0;
+}
+
+Real
+PolycrystalEBSD::getVariableValue(unsigned int op_index, const Point & p) const
+{
+  std::vector<unsigned int> grain_ids;
+  getGrainsBasedOnPoint(p, grain_ids);
+
+  if (grain_ids.empty())
+    return -1.0;
```

Comment:
No you hit it in a multiphase problem. Well... not with this PR but the follow on one where I'm getting the phase stuff all working properly again before trying to use this new object with any of the phase problems. It turns out that you want to build one of these objects per phase instead of trying to read in all the phases in one object. It greatly simplifies the logic in the object. In those cases you might hit a place where there are no active grains on a phase.